File: export-slides.js
// Export carousel slides as 1080×1440 PNGs
// Usage: node export-slides.js carousels/<slug>/index.html

const { chromium } = require('playwright');
const path = require('path');
const fs = require('fs');

const SLIDE_W = 420;
const SLIDE_H = 560;
const EXPORT_W = 1080;
const EXPORT_H = 1440;
const SCALE = EXPORT_W / SLIDE_W; // 2.5714…

async function exportSlides(htmlFile) {
  const absolutePath = path.resolve(htmlFile);

  if (!fs.existsSync(absolutePath)) {
    console.error(`File not found: ${absolutePath}`);
    process.exit(1);
  }

  const slidesDir = path.join(path.dirname(absolutePath), 'slides');
  if (!fs.existsSync(slidesDir)) fs.mkdirSync(slidesDir, { recursive: true });

  const browser = await chromium.launch();
  const context = await browser.newContext({
    viewport: { width: SLIDE_W, height: SLIDE_H },
    deviceScaleFactor: SCALE,
  });
  const page = await context.newPage();

  await page.goto(`file://${absolutePath}`);
  await page.waitForLoadState('networkidle');
  await page.waitForTimeout(800); // let fonts paint

  const slideCount = await page.evaluate(
    () => document.querySelectorAll('[data-slide]').length
  );

  if (slideCount === 0) {
    console.error('No [data-slide] elements found. Check the HTML template.');
    await browser.close();
    process.exit(1);
  }

  console.log(`\nExporting ${slideCount} slides → ${EXPORT_W}×${EXPORT_H}px`);
  console.log(`Output: ${slidesDir}\n`);

  for (let i = 0; i < slideCount; i++) {
    await page.evaluate((idx) => window.goToSlide(idx), i);
    await page.waitForTimeout(120);

    const slideNum = String(i + 1).padStart(2, '0');
    const outPath = path.join(slidesDir, `slide-${slideNum}.png`);

    await page.screenshot({
      path: outPath,
      clip: { x: 0, y: 0, width: SLIDE_W, height: SLIDE_H },
    });

    console.log(`  ✓ slide-${slideNum}.png`);
  }

  await browser.close();
  console.log(`\nDone. Open ${slidesDir} to grab your PNGs.\n`);
}

const file = process.argv[2];
if (!file) {
  console.error('Usage: node export-slides.js <path/to/carousel/index.html>');
  process.exit(1);
}

exportSlides(file).catch((err) => {
  console.error(err);
  process.exit(1);
});

File: export-slides.py
#!/usr/bin/env python3
# Export carousel slides as 1080×1440 PNGs (Python / Playwright fallback)
# Usage: python3 export-slides.py carousels/<slug>/index.html
#
# Install once:
#   pip3 install playwright          (add --break-system-packages if pip refuses)
#   python3 -m playwright install chromium

import sys
import os
from pathlib import Path

try:
    from playwright.sync_api import sync_playwright
except ImportError:
    print("Playwright not found. Run: pip3 install playwright && python3 -m playwright install chromium")
    sys.exit(1)

SLIDE_W, SLIDE_H = 420, 560
EXPORT_W, EXPORT_H = 1080, 1440
SCALE = EXPORT_W / SLIDE_W  # 2.5714…


def export_slides(html_file: str):
    html_path = Path(html_file).resolve()

    if not html_path.exists():
        print(f"File not found: {html_path}")
        sys.exit(1)

    slides_dir = html_path.parent / "slides"
    slides_dir.mkdir(parents=True, exist_ok=True)

    with sync_playwright() as p:
        browser = p.chromium.launch()
        context = browser.new_context(
            viewport={"width": SLIDE_W, "height": SLIDE_H},
            device_scale_factor=SCALE,
        )
        page = context.new_page()
        page.goto(f"file://{html_path}")
        page.wait_for_load_state("networkidle")
        page.wait_for_timeout(800)

        slide_count = page.evaluate(
            "document.querySelectorAll('[data-slide]').length"
        )

        if slide_count == 0:
            print("No [data-slide] elements found. Check the HTML template.")
            browser.close()
            sys.exit(1)

        print(f"\nExporting {slide_count} slides → {EXPORT_W}×{EXPORT_H}px")
        print(f"Output: {slides_dir}\n")

        for i in range(slide_count):
            page.evaluate(f"window.goToSlide({i})")
            page.wait_for_timeout(120)

            slide_num = str(i + 1).zfill(2)
            out_path = str(slides_dir / f"slide-{slide_num}.png")

            page.screenshot(
                path=out_path,
                clip={"x": 0, "y": 0, "width": SLIDE_W, "height": SLIDE_H},
            )
            print(f"  ✓ slide-{slide_num}.png")

        browser.close()
        print(f"\nDone. Open {slides_dir} to grab your PNGs.\n")


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python3 export-slides.py <path/to/carousel/index.html>")
        sys.exit(1)
    export_slides(sys.argv[1])
